evaluator/functions.go

```diff
@@ -1319,3 +1346,57 @@ func sleepFn(tok token.Token, args ...object.Object) object.Object {
 
 	return NULL
 }
+
+// source("fileName")
+// aka src(), include(), import(), require()
+var sourceMaxRecursionDepth = 5
+
+func sourceFn(tok token.Token, args ...object.Object) object.Object {
+	// limit source file recursion depth
+	if sourceMaxRecursionDepth <= 0 {
+		return newError(tok, "maximum source file recursion depth exceeded")
+	}
+	// mark this source recursion level
+	sourceMaxRecursionDepth--
+
+	err := validateArgs(tok, "source", args, 1, [][]string{{object.STRING_OBJ}})
+	if err != nil {
+		return err
+	}
+	// load the source file
+	fileName, _ := util.ExpandPath(args[0].Inspect())
+	code, error := ioutil.ReadFile(fileName)
+	if error != nil {
+		// source file path does not exist
+		return newError(tok, "source file not found: %s", fileName)
+	}
+	// parse it
+	l := lexer.New(string(code))
+	p := parser.New(l)
+	program := p.ParseProgram()
+	errors := p.Errors()
+	if len(errors) != 0 {
+		errMsg := fmt.Sprintf("%s", " parser errors:\n")
+		for _, msg := range errors {
+			errMsg += fmt.Sprintf("%s", "\t"+msg+"\n")
+		}
+		return newError(tok, "source file: %s\n%s", fileName, errMsg)
+	}
+	// invoke BeginEval() passing in the sourced program, globalEnv, and our lexer
+	// we save the current global lexer and restore it after we return from BeginEval()
+	savedLexer := lex
```

Comment:
why are we saving the lexer aside like this? I don't see us using the `lex` variable at all

Reply:
`lex` is the global lexer var that is used whenever error line reporting occurs. We have to save and restore it because each source file has its own lexer that is passed to `BeginEval(prog, env, lexer)` which then sets the global `lex` to the arg lexer for its program. This allows error line reporting to be relative to the sourced file when we have nested includes. The alternative is to clutter the interfaces of every function in the evaluator package by passing the current lexer that is only used for error line reporting.

I have added some explanatory comments to the `source()` func.